app/models/watchlist_instance.rb

```diff
@@ -83,4 +115,176 @@ def resolve_watchlist_instance
     end
   end
 
+private
+  
+  def self.add_new_instances_for_conditions(conditions, course)
+    new_instances = []
+    course_user_data = CourseUserDatum.where(course_id: course.id, instructor: false, course_assistant: false)
+    # HUGE TRANSACTION AHEAD
+    ActiveRecord::Base.transaction do
+      conditions.each do |condition|
+        case condition.condition_type
+        
+        when "grace_day_usage"
+          grace_day_threshold = condition.parameters[:grace_day_threshold]
+          date = condition.parameters[:date]
+          asmts = course.assessments.ordered
+          for cud in course_user_data
+            asmts_before_date = asmts.where("updated_at < ?", date)
+            latest_asmt = asmts_before_date.last
+            next if latest_asmt.nil?
+            new_instance = self.add_new_instance_for_cud_grace_day_usage(course, condition.id, cud, asmts_before_date, grace_day_threshold)
```

Comment:
should typecheck and convert grace_day_threshold to integers, else the comparison will fail later